src/wpilib/i2c.rs

```diff
@@ -0,0 +1,134 @@
+            0 => Ok(I2cInterface { port: p, device_address: addr }),
+            _ => Err(HalError(0)),
+        }
+    }
+
+    /// perform a simultaneous read from and write to an i2c device
+    pub fn transaction(&mut self, sent: &[u8], received: &mut [u8]) -> Result<(), I2cError> {
+        let status = unsafe {
+            HAL_TransactionI2C(self.port as i32, self.device_address, sent.as_ptr() as *mut u8, sent.len() as i32,
+                                                                      received.as_mut_ptr(), received.len() as i32)
+        };
+        match status {
+            -1 => Err(I2cError::TransferAbort),
+            x if x >= 0 => Ok(()),
+            _ => Err(I2cError::IOError),
+        }
+    }
+
+    /// reads received message to inputed byte slice
+    pub fn read(&self, received: &mut [u8]) -> Result<(), I2cError> {
+        let status = unsafe {
+            HAL_ReadI2C(self.port as i32, self.device_address, received.as_mut_ptr(), received.len() as i32)
+        };
+        match status {
+            -1 => Err(I2cError::TransferAbort),
+            x if x >= 0 => Ok(()),
+            _ => Err(I2cError::IOError),
+        }
+    }
+
+    /// writes byte slice to i2c device
+    pub fn write(&mut self, sent: &[u8]) -> Result<(), I2cError> {
+        let status = unsafe {
+            HAL_WriteI2C(self.port as i32, self.device_address, sent.as_ptr() as *mut u8, sent.len() as i32)
+        };
+        match status {
+            -1 => Err(I2cError::TransferAbort),
+            x if x >= 0 => Ok(()),
+            _ => Err(I2cError::IOError),
+        }
+    }
+
+    /// performs string transaction with connected i2c device
+    pub fn string_transaction(&mut self, sent_string: &str, received_string: &str) -> Result<(), I2cError> {
+        let status = unsafe {
+            HAL_TransactionI2C(self.port as i32, self.device_address, sent_string.as_bytes().as_ptr() as *mut u8, sent_string.len() as i32,
+                                                                      received_string.as_bytes().as_ptr() as *mut u8, received_string.len() as i32)
+        };
+        match status {
+            -1 => Err(I2cError::TransferAbort),
+            x if x >= 0 => Ok(()),
+            _ => Err(I2cError::IOError),
+        }
+    }
+
+    /// reads received message to str
+    pub fn read_string(&self, received_string: &mut str) -> Result<(), I2cError> {
+        let status = unsafe {
+            HAL_ReadI2C(self.port as i32, self.device_address, received_string.as_bytes().as_ptr() as *mut u8, received_string.len() as i32)
+        };
+        match status {
+            -1 => Err(I2cError::TransferAbort),
+            //check to make sure a valid string was received
+            x if x >= 0 => match unsafe { String::from_utf8(slice::from_raw_parts(received_string.as_bytes().as_ptr(),
+                                                                                  received_string.len()).to_vec()) } {
```

Comment:
This whole thing does a bunch of allocation purely to check that the length of a slice is valid. It could really be resolved by checking that `received_string.len() == x`.